### PR TITLE
feat: Initial Implementation of Storager

### PR DESCRIPTION
**Pull request by abyss-w**
feat: Initial Implementation of Storager

### CODE REVIEW

File: service.toml [hunk 1]
```diff
@@ -1,6 +1,11 @@
 name = "azfile"
 
 [namespace.storage]
+features = ["virtual_dir"]
```
Comment (JinnyYi) [inline service.toml]: `dir` is natively supported in azfile. We can implement `Direr` with the public APIs instead of simulate it.

Comment (abyss-w) [reply service.toml]: Got it!

File: storage.go [hunk 2]
```diff
@@ -2,35 +2,283 @@ package azfile
...
+func (s *Storage) createDir(ctx context.Context, path string, opt pairStorageCreateDir) (o *Object, err error) {
+	rp := s.getAbsPath(path)
+
+	attribute := azfile.FileAttributeNone
+
+	properties := azfile.SMBProperties{
+		FileAttributes: &attribute,
+	}
+
+	_, err = s.client.NewDirectoryURL(path).Create(ctx, nil, properties)
```
Comment (JinnyYi) [inline storage.go]: Create directory doesn't support `overwrite`, so we need to check whether the dir exists here.
ref: https://docs.microsoft.com/en-us/rest/api/storageservices/create-directory#remarks

File: storage.go [hunk 3]
```diff
@@ -2,35 +2,283 @@ package azfile
...
 
 func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
-	panic("not implemented")
+	if opt.HasIoCallback {
+		r = iowrap.CallbackReader(r, opt.IoCallback)
+	}
+
+	body := iowrap.SizedReadSeekCloser(r, size)
+
+	_, err = s.client.NewFileURL(path).UploadRange(ctx, 0, body, nil)
```
Comment (JinnyYi) [inline storage.go]: I'm not sure whether we can use `UploadRange` to create a new file as the `Put Range` operation writes a range of byte to a file. This operation can only be called on an existing file.
ref: https://docs.microsoft.com/en-us/rest/api/storageservices/put-range#remarks

Comment (abyss-w) [reply storage.go]: Maybe we should use `Create`.

File: storage.go [hunk 4]
```diff
@@ -273,9 +288,27 @@ func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int6
...
+
+	// `Create` only initializes the file.
+	// ref: https://docs.microsoft.com/en-us/rest/api/storageservices/create-file
+	_, err = s.client.NewFileURL(path).Create(ctx, size, headers, nil)
+	if err != nil {
+		return 0, err
+	}
+
 	body := iowrap.SizedReadSeekCloser(r, size)
+	transactionalMD5, err := base64.StdEncoding.DecodeString(opt.ContentMd5)
```
Comment (JinnyYi) [inline storage.go]: `content_md5` is an optional pair, we should check `HasContentMd5` first.

Comment (abyss-w) [reply storage.go]: Got it!

File: storage.go [hunk 5]
```diff
@@ -2,35 +2,316 @@ package azfile
...
+
+	properties := azfile.SMBProperties{
+		FileAttributes: &attribute,
+	}
+
+	_, err = s.client.NewDirectoryURL(path).GetProperties(ctx)
+	if err == nil {
+		// The directory exist, we should delete it.
+		// ref: https://github.com/beyondstorage/specs/blob/master/rfcs/134-write-behavior-consistency.md
+		_, err = s.client.NewDirectoryURL(path).Delete(ctx)
```
Comment (JinnyYi) [inline storage.go]: We don't need to delete the existing dir and recreate here. We can set the metadata for the existing dir object.

File: storage.go [hunk 6]
```diff
@@ -2,35 +2,316 @@ package azfile
...
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.WorkDir = s.workDir
+	return meta
+}
+
+func (s *Storage) nextObjectPageByDir(ctx context.Context, page *ObjectPage) error {
```
Comment (JinnyYi) [inline storage.go]: Looks the same with `nextObjectPageByPrefix`?

File: storage.go [hunk 7]
```diff
@@ -41,15 +40,15 @@ func (s *Storage) createDir(ctx context.Context, path string, opt pairStorageCre
...
 	_, err = s.client.NewDirectoryURL(path).GetProperties(ctx)
 	if err == nil {
-		// The directory exist, we should delete it.
-		// ref: https://github.com/beyondstorage/specs/blob/master/rfcs/134-write-behavior-consistency.md
-		_, err = s.client.NewDirectoryURL(path).Delete(ctx)
-		if err != nil {
-			return nil, err
-		}
+		// The directory exist, we should reset the metadata.
+		_, err = s.client.NewDirectoryURL(path).SetMetadata(ctx, nil)
```
Comment (JinnyYi) [inline storage.go]: Sorry for my inaccurate expression. We don't need to call `SetMetadata` here. We can use the metadata returned by `GetProperties` like `lastModified` to assign the value for `o.lastModified`. Like what we did in https://github.com/beyondstorage/go-service-kodo/blob/master/storage.go#L51-L60

File: storage.go [hunk 8]
```diff
@@ -2,35 +2,266 @@ package azfile
...
+		o.SetLastModified(fileOutput.LastModified())
+
+		if v := string(fileOutput.ETag()); v != "" {
+			o.SetEtag(v)
+		}
+		if v := fileOutput.ContentType(); v != "" {
+			o.SetContentType(v)
+		}
+		if v := fileOutput.ContentMD5(); len(v) > 0 {
+			o.SetContentMd5(base64.StdEncoding.EncodeToString(v))
```
Comment (JinnyYi) [inline storage.go:221]: ping @Xuanwo I'm not sure whether we need base64 encode here? Similar question in https://github.com/beyondstorage/go-service-azfile/pull/2/files#diff-8127f6766ba5110c7a99d07305f11d18295fcb2cb08dddb105ab7d55898742c7R254.

**Comment by Xuanwo** [comment]
This week is our 1st cleanup week, we will resume the work here in the next week: https://forum.beyondstorage.io/t/1st-cleanup-week-for-go-storage/213/12

<blockquote><div> BeyondStorage</div><div><strong><a href="https://forum.beyondstorage.io/t/1st-cleanup-week-for-go-storage/213/12">1st Cleanup Week for go-storage</a></strong></div><div>Hello everyone, we will start a cleanup week from 9/6 to 9/10.  In next week, we will stop any new feature development of go-storage and focus on bugfix, docs like:   every project’s README.md update (update links of matrix room, integration tests, and so on) go-storage-example enhancement. go-storage user guide and developer guide. Clean up issues we are working on and plan to do. File new issues about any problems that meet (could be sovled later) …  Next week, all our maintainers and committe...</div></blockquote>

File: storage.go (hunk 8 above)
Comment (Xuanwo) [reply storage.go:221]: Yes, our content md5 is based encoded, maybe we need to add description in `content_md5`.

File: storage.go [hunk 9]
```diff
@@ -2,35 +2,268 @@ package azfile
...
+	var transactionalMD5 []byte
+	if opt.HasContentMd5 {
+		transactionalMD5, err = base64.StdEncoding.DecodeString(opt.ContentMd5)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// Since `Create' only initializes the file, we need to call `UploadRange' to write the contents to the file.
+	_, err = s.client.NewFileURL(path).UploadRange(ctx, 0, body, transactionalMD5)
```
Comment (Xuanwo) [inline storage.go:263]: What happened if `transactionalMD5` is empty?

Comment (abyss-w) [reply storage.go:263]: This is an optional value, so the null value does not affect the API call.

**Comment by Xuanwo** [comment]
Most LGTM.

**Pull request merged**